File: ai/openai_chat.go
// OpenAI 相容 Chat Completions：供玩家對 NPC Talk 走雲端／Web API（/v1/chat/completions）。
package ai

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"strings"
	"time"
)

// CallOpenAICompatibleChat 呼叫 baseURL/v1/chat/completions（baseURL 須為 …/v1 結尾，例如 https://api.openai.com/v1）。
// apiKey 可空（部分自建閘道不需）；非空則帶 Authorization: Bearer。
func CallOpenAICompatibleChat(baseURL, apiKey, model, systemPrompt, userMessage string) (string, error) {
	baseURL = strings.TrimSpace(strings.TrimSuffix(baseURL, "/"))
	if baseURL == "" || strings.TrimSpace(model) == "" {
		return "", errors.New("player talk api not configured")
	}
	endpoint := baseURL + "/chat/completions"

	body := map[string]interface{}{
		"model": model,
		"messages": []map[string]string{
			{"role": "system", "content": systemPrompt},
			{"role": "user", "content": userMessage},
		},
		"temperature": 0.82,
		"max_tokens":  200,
	}
	// OpenRouter：部分模型會做 reasoning／thinking，拖長首字延遲；關閉 effort 可接近「張嘴就答」的 Flash 體感（見 openrouter.ai docs reasoning-tokens）。
	if strings.Contains(strings.ToLower(baseURL), "openrouter.ai") {
		body["reasoning"] = map[string]interface{}{
			"effort": "none",
		}
	}
	raw, err := json.Marshal(body)
	if err != nil {
		return "", err
	}
	req, err := http.NewRequest(http.MethodPost, endpoint, bytes.NewReader(raw))
	if err != nil {
		return "", err
	}
	req.Header.Set("Content-Type", "application/json")
	if strings.TrimSpace(apiKey) != "" {
		req.Header.Set("Authorization", "Bearer "+strings.TrimSpace(apiKey))
	}

	client := &http.Client{Timeout: 90 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("chat api %s", resp.Status)
	}

	var out struct {
		Choices []struct {
			Message struct {
				Content string `json:"content"`
			} `json:"message"`
		} `json:"choices"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&out); err != nil {
		return "", err
	}
	if len(out.Choices) == 0 {
		return "", errors.New("empty choices")
	}
	reply := strings.TrimSpace(out.Choices[0].Message.Content)
	if reply == "" {
		return "", errors.New("empty reply")
	}
	gopen, gclose := string(rune(0x300c)), string(rune(0x300d))
	if strings.HasPrefix(reply, gopen) && strings.Contains(reply, gclose) {
		if i := strings.Index(reply, gclose); i > 0 {
			reply = reply[len(gopen):i]
		}
	}
	return reply, nil
}

File: src/ai/openai_chat.rs
// OpenAI 相容 Chat Completions，對齊 Go ai/openai_chat.go。

use serde::Deserialize;
use serde_json::json;
use std::time::Duration;

use super::talk::strip_reply_guillemet_outer;

#[derive(Debug, Deserialize)]
struct OpenAiChatResponse {
    choices: Vec<OpenAiChoice>,
}

#[derive(Debug, Deserialize)]
struct OpenAiChoice {
    message: OpenAiMsg,
}

#[derive(Debug, Deserialize)]
struct OpenAiMsg {
    content: String,
}

/// 呼叫 `baseURL/chat/completions`（`baseURL` 須為 `…/v1`，例如 `https://api.openai.com/v1`）。
pub fn call_openai_compatible_chat(
    base_url: &str,
    api_key: &str,
    model: &str,
    system_prompt: &str,
    user_message: &str,
) -> anyhow::Result<String> {
    let base_url = base_url.trim().trim_end_matches('/');
    let model = model.trim();
    if base_url.is_empty() || model.is_empty() {
        anyhow::bail!("player talk api not configured");
    }
    let endpoint = format!("{base_url}/chat/completions");
    let mut body = json!({
        "model": model,
        "messages": [
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": user_message},
        ],
        "temperature": 0.82,
        "max_tokens": 200,
    });
    if base_url.to_lowercase().contains("openrouter.ai") {
        body.as_object_mut()
            .expect("object")
            .insert(
                "reasoning".to_string(),
                json!({ "effort": "none" }),
            );
    }
    let client = reqwest::blocking::Client::builder()
        .timeout(Duration::from_secs(90))
        .build()?;
    let mut req = client.post(&endpoint).json(&body);
    let key = api_key.trim();
    if !key.is_empty() {
        req = req.header("Authorization", format!("Bearer {key}"));
    }
    let resp = req.send()?;
    if !resp.status().is_success() {
        anyhow::bail!("chat api {}", resp.status());
    }
    let out: OpenAiChatResponse = resp.json()?;
    let choice = out
        .choices
        .first()
        .ok_or_else(|| anyhow::anyhow!("empty choices"))?;
    let reply = choice.message.content.trim();
    if reply.is_empty() {
        anyhow::bail!("empty reply");
    }
    Ok(strip_reply_guillemet_outer(reply))
}
